pkg/resource/stack/secrets.go

```diff
@@ -86,171 +89,507 @@ func (s NamedStackSecretsProvider) OfType(ty string, state json.RawMessage) (sec
+//	secret := &resource.Secret{}
+//	cache.Write("plaintext", "ciphertext", secret)
+//	plaintext, ok := cache.TryDecrypt("ciphertext") // "plaintext", true
+//	ciphertext, ok := cache.TryEncrypt(secret, "plaintext") // "ciphertext", true
+type SecretCache interface {
+	// Write stores the plaintext, ciphertext, and secret in the cache, overwriting any previous entry for the secret.
+	Write(plaintext, ciphertext string, secret *resource.Secret)
+	// LookupCiphertext returns the cached ciphertext for the given secret and plaintext, if it exists.
+	LookupCiphertext(secret *resource.Secret, plaintext string) (string, bool)
+	// LookupPlaintext returns the cached plaintext for the given ciphertext, if it exists.
+	LookupPlaintext(ciphertext string) (string, bool)
 }
 
-// encryptSecret encrypts the plaintext associated with the given secret value.
-func (csm *cachingSecretsManager) encryptSecret(ctx context.Context,
-	secret *resource.Secret, plaintext string,
-) (string, error) {
-	// If the cache has an entry for this secret and the plaintext has not changed, re-use the ciphertext.
-	//
-	// Otherwise, re-encrypt the plaintext and update the cache.
-	entry, ok := csm.cache[secret]
-	if ok && entry.plaintext == plaintext {
-		return entry.ciphertext, nil
+type nullSecretCache struct{}
+
+func (nullSecretCache) Write(plaintext, ciphertext string, secret *resource.Secret) {}
+func (nullSecretCache) LookupCiphertext(secret *resource.Secret, plaintext string) (string, bool) {
+	return "", false
+}
+
+func (nullSecretCache) LookupPlaintext(ciphertext string) (string, bool) {
+	return "", false
+}
+
+type secretCache struct {
+	bySecret     sync.Map
+	byCiphertext sync.Map
+}
+
+type secretCacheEntry struct {
+	plaintext  string
+	ciphertext string
+	secret     *resource.Secret
+}
+
+// NewSecretCache returns a new secretCache which allows the bidirectional cached conversion between:
+// `ciphertext <-> plaintext + secret pointer`. The same plaintext can be associated with multiple secrets, each of
+// which will have their own ciphertexts which should not be shared.
+//
+// All methods are thread-safe and can be called concurrently by multiple goroutines.
+// The cache can be disabled by setting the environment variable PULUMI_DISABLE_SECRET_CACHE to "true".
+//
+//	cache := NewSecretCache()
+//	secret := &resource.Secret{}
+//	cache.Write("plaintext", "ciphertext", secret)
+//	plaintext, ok := cache.TryDecrypt("ciphertext") // "plaintext", true
+//	ciphertext, ok := cache.TryEncrypt(secret, "plaintext") // "ciphertext", true
+func NewSecretCache() SecretCache {
+	// If the environment variable PULUMI_DISABLE_SECRET_CACHE is set to "true", the secret cache will be disabled and
+	// no entries will be stored. This is a short-term escape hatch in case there's unforeseen issues with expanded
+	// caching scopes and should be removable once we're confident that customers are not affected.
+	if os.Getenv("PULUMI_DISABLE_SECRET_CACHE") == "true" {
```

Comment:
Wonder if we should add this in `sdk/go/common/env/env.go` so that it appears in `pulumi about env`/fits in with other env vars we have (there are a few exceptions/ones we've missed but for the most part we are quite good about this).

Reply:
TIL